File: Foundations/Projects/BINARY/binary_fragmentation/core/provenance.py
"""
core/provenance.py
==================
First-Class Provenance Tracking and Rollback Ledger.
Enforces that every transformation records:
- State ID & Parent State ID
- Operator name & Parameters
- Information removed & Information added
- Checksums and Reversibility status
"""

from __future__ import annotations
import copy
import hashlib
import time
from dataclasses import dataclass, field, asdict
from typing import Any, Dict, List, Optional, Tuple


@dataclass
class ProvenanceRecord:
    """Detailed causal ledger entry documenting a single state transformation."""
    state_id: str
    parent_state_id: Optional[str]
    operator: str
    timestamp: float = field(default_factory=time.time)
    parameters: Dict[str, Any] = field(default_factory=dict)
    info_removed: Dict[str, Any] = field(default_factory=dict)
    info_added: Dict[str, Any] = field(default_factory=dict)
    checksum_before: str = ""
    checksum_after: str = ""
    reversible: bool = True
    reversibility_notes: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> ProvenanceRecord:
        return cls(
            state_id=str(data["state_id"]),
            parent_state_id=data.get("parent_state_id"),
            operator=str(data.get("operator", "unknown")),
            timestamp=float(data.get("timestamp", time.time())),
            parameters=data.get("parameters", {}),
            info_removed=data.get("info_removed", {}),
            info_added=data.get("info_added", {}),
            checksum_before=str(data.get("checksum_before", "")),
            checksum_after=str(data.get("checksum_after", "")),
            reversible=bool(data.get("reversible", True)),
            reversibility_notes=str(data.get("reversibility_notes", "")),
        )


class ProvenanceLedger:
    """
    Append-only causal lineage tracking ledger for state evolutions.
    Maintains full chain of custody and evaluates causal graph depth.
    """

    def __init__(self) -> None:
        self.records: List[ProvenanceRecord] = []
        self._index_by_state: Dict[str, ProvenanceRecord] = {}

    def append(self, record: ProvenanceRecord) -> None:
        self.records.append(record)
        self._index_by_state[record.state_id] = record

    def get_history(self, state_id: str) -> List[ProvenanceRecord]:
        """Traces the backward causal ancestry from a given state to root."""
        chain: List[ProvenanceRecord] = []
        current_id: Optional[str] = state_id
        visited = set()

        while current_id and current_id in self._index_by_state:
            if current_id in visited:
                break  # Prevent infinite cycle
            visited.add(current_id)
            rec = self._index_by_state[current_id]
            chain.append(rec)
            current_id = rec.parent_state_id

        chain.reverse()
        return chain

    def compute_irreversibility_score(self) -> float:
        """
        Fraction of transformations marked irreversible in the ledger.
        Range: 0.0 (fully reversible) to 1.0 (completely irreversible).
        """
        if not self.records:
            return 0.0
        irreversible_count = sum(1 for r in self.records if not r.reversible)
        return irreversible_count / len(self.records)

    def to_dict_list(self) -> List[Dict[str, Any]]:
        return [r.to_dict() for r in self.records]

File: packages/rust/binary-fragmentation/src/provenance.rs
//! `core/provenance.py` — append-only causal lineage tracking.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ProvenanceRecord {
    pub state_id: String,
    pub parent_state_id: Option<String>,
    pub operator: String,
    pub timestamp: f64,
    pub parameters: Map<String, Value>,
    pub info_removed: Map<String, Value>,
    pub info_added: Map<String, Value>,
    pub checksum_before: String,
    pub checksum_after: String,
    pub reversible: bool,
    pub reversibility_notes: String,
}

impl ProvenanceRecord {
    pub fn new(
        state_id: impl Into<String>,
        parent_state_id: Option<String>,
        operator: impl Into<String>,
        parameters: Map<String, Value>,
        info_removed: Map<String, Value>,
        info_added: Map<String, Value>,
        checksum_before: impl Into<String>,
        checksum_after: impl Into<String>,
        reversible: bool,
        reversibility_notes: impl Into<String>,
    ) -> Self {
        Self {
            state_id: state_id.into(),
            parent_state_id,
            operator: operator.into(),
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_secs_f64())
                .unwrap_or(0.0),
            parameters,
            info_removed,
            info_added,
            checksum_before: checksum_before.into(),
            checksum_after: checksum_after.into(),
            reversible,
            reversibility_notes: reversibility_notes.into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ProvenanceLedger {
    pub records: Vec<ProvenanceRecord>,
}

impl ProvenanceLedger {
    pub fn new() -> Self {
        Self {
            records: Vec::new(),
        }
    }

    pub fn append(&mut self, record: ProvenanceRecord) {
        self.records.push(record);
    }

    /// Trace the causal chain from `state_id` back to the root, newest first.
    pub fn get_history(&self, state_id: &str) -> Vec<ProvenanceRecord> {
        let mut history = Vec::new();
        let mut current: Option<String> = Some(state_id.to_string());
        while let Some(cid) = current {
            let found = self.records.iter().find(|r| r.state_id == cid);
            match found {
                Some(record) => {
                    history.push(record.clone());
                    current = record.parent_state_id.clone();
                }
                None => break,
            }
        }
        history
    }

    /// Fraction of irreversible records (0.0 if no records).
    pub fn compute_irreversibility_score(&self) -> f64 {
        if self.records.is_empty() {
            return 0.0;
        }
        let irreversible = self.records.iter().filter(|r| !r.reversible).count();
        irreversible as f64 / self.records.len() as f64
    }

    pub fn to_dict_list(&self) -> Vec<Value> {
        self.records
            .iter()
            .map(|r| serde_json::to_value(r).expect("serialize"))
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn record(id: &str, parent: Option<&str>, reversible: bool) -> ProvenanceRecord {
        ProvenanceRecord::new(
            id,
            parent.map(str::to_string),
            "Op",
            Map::new(),
            Map::new(),
            Map::new(),
            "b",
            "a",
            reversible,
            "",
        )
    }

    #[test]
    fn history_traces_backward_chain() {
        let mut ledger = ProvenanceLedger::new();
        ledger.append(record("s1", None, true));
        ledger.append(record("s2", Some("s1"), true));
        ledger.append(record("s3", Some("s2"), false));

        let history = ledger.get_history("s3");
        assert_eq!(history.len(), 3);
        assert_eq!(history[0].state_id, "s3");
        assert_eq!(history[1].state_id, "s2");
        assert_eq!(history[2].state_id, "s1");
    }

    #[test]
    fn irreversibility_score_is_fraction() {
        let mut ledger = ProvenanceLedger::new();
        assert_eq!(ledger.compute_irreversibility_score(), 0.0);
        ledger.append(record("s1", None, true));
        ledger.append(record("s2", Some("s1"), false));
        assert!((ledger.compute_irreversibility_score() - 0.5).abs() < 1e-12);
    }
}